### PR TITLE
Connection Gating in a libp2p node

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For #872 .

Core PR at https://github.com/libp2p/go-libp2p-core/pull/139.
Swarm PR at https://github.com/libp2p/go-libp2p-swarm/pull/201.
Transport Upgrader PR at https://github.com/libp2p/go-libp2p-transport-upgrader/pull/55.
MultiAddr Filter PR at https://github.com/libp2p/go-maddr-filter/pull/21.

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -4,7 +4,6 @@ import (
 	"context"
 	"crypto/rand"
 	"fmt"
-
```
Comment (willscott) [inline config/config.go:7]: the import block separation is desirable, i think

File: options.go [hunk 2]
```diff
@@ -5,22 +5,20 @@ package libp2p
...
-	"net"
-
 	"github.com/libp2p/go-libp2p-core/connmgr"
 	"github.com/libp2p/go-libp2p-core/crypto"
 	"github.com/libp2p/go-libp2p-core/metrics"
 	"github.com/libp2p/go-libp2p-core/network"
 	"github.com/libp2p/go-libp2p-core/peer"
 	"github.com/libp2p/go-libp2p-core/peerstore"
 	"github.com/libp2p/go-libp2p-core/pnet"
+	"github.com/pkg/errors"
```
Comment (willscott) [inline options.go:15]: you don't need this dep. either the system built-in `"errors"` can be used for `errors.New`, or you can use the already included `fmt` for `fmt.Errorf()`.

File: options.go [hunk 3]
```diff
@@ -302,28 +300,16 @@ func EnableNATService() Option {
...
 }
 
-// FilterAddresses configures libp2p to never dial nor accept connections from
-// the given addresses. FilterAddresses should be used for cases where the
-// addresses you want to deny are known ahead of time.
-func FilterAddresses(addrs ...*net.IPNet) Option {
+// ConnectionGater configures libp2p to use the given Connection Gater
+// to reject incoming/outgoing connections based om the logic implemented in the gater.
+// Note: This Option has replaced the earlier "Filters" AND "FilterAddresses" Options.
+// Please use "filter.ToConnectionGater" to configure this of you already have an instance of a "go-maddr-filter.Filter".
```
Comment (willscott) [inline options.go:306]:
```suggestion
// Existing users of go-maddr-filter.Filter may use "filter.ToConnectionGater" for compatibility.
```

**Comment by aarshkshah1992** [comment]
Closing in favour of #881 .

**Pull request closed without merging**